### ISSUE TITLE
ERROR: Unable to detect GNU compiler type: cc1: error: /dev/null/usr/include: Not a directory

### ISSUE THREAD

**Issue by haata** [mesonbuild/meson#6482]
**Describe the bug**

Using gcc 4.8.4 compiled for the VMWare nvdk, meson throws an exception trying to detect the compiler. (I don't have a way of rebuilding this compiler FWIW)
I believe due to attempts at toolchain isolation, VMWare has set the sysroot to /dev/null. I believe this is the case for all of VMWare's SDKs containing gcc.

I think `--with-sysroot=/dev/null` is the offending option.
```
/usr/lib/gcc/x86_64-vmk-linux-gnu/4.8.4/plugin/include/configargs.h:2:static const char configuration_arguments[] = "/build/mts/
release/bora-3656972/cayman_esx_gcc/gcc/src/configure --build=x86_64-unknown-linux-gnu --host=x86_64-unknown-linux-gnu --target
=x86_64-vmk-linux-gnu --with-sysroot=/dev/null --with-build-sysroot=/build/mts/release/bora-3656972/cayman_esx_gcc/build/sysroo
t --prefix=/usr --enable-languages=c,c++ --disable-sjlj-exceptions --enable-frame-pointer";
```
I can't find any way to force meson to use an alternate sysroot during compiler detection without modifying meson itself.

**To Reproduce**

Retrieve ESX NVDK 6.5, and build a very simple meson file using the installed gcc. It fails at the project definition.

**Expected behavior**
In the case of badly configured compilers there should be a way to set an alternate sysroot (I do have one for the SDK). Or at least skip this check altogether.

**system parameters**
* Cross build
* Linux CentOS 7
* Python 3.6.8
* meson 0.53.0
* ninja 1.7.2

**Comment by haata** [comment]
**Meson Default**
```
/usr/bin/gcc -E -dM -
cc1: error: /dev/null/usr/include: Not a directory
```

**Working Version**
```
/usr/bin/gcc --sysroot=--sysroot=/opt/vmware/toolchain/cayman_esx_glibc-7fc626201c38c3a7e2472179c969a565b101f7ba/sysroot -E -dM -
```

So my current options at the moment (also as assistance to anyone who runs into this with other compilers with an invalid sysroot).

1. Patch this line to include a `--sysroot=/opt/vmware/toolchain/cayman_esx_glibc-7fc626201c38c3a7e2472179c969a565b101f7ba/sysroot` using some sort of cross file variable (or native file). The sysroot changes depending on which version of vmware is being targeted.

2. Write a wrapper script that adds arguments to every invocation of gcc. For now this is probably what I'll have to do in order to make any progress.

3. Patch the compiler binary to remove the /dev/null.

4. Patch the running version of meson to add in the sysroot argument.

**Comment by dcbaker** [comment]
If you add:
```ini
[properties]
c_args = ['--sysroot=--sysroot=/opt/vmware/toolchain/cayman_esx_glibc-7fc626201c38c3a7e2472179c969a565b101f7ba/sysroot']
```

to your cross file, does that resolve the issue?

**Comment by haata** [comment]
I don't believe c_args is passed for this check: https://github.com/mesonbuild/meson/blob/master/mesonbuild/environment.py#L668
```c
args = compiler + ['-E', '-dM', '-']
```

(I tried it, didn't work, same error).

**Comment by haata** [comment]
For those that run into this I currently use a wrapper script for both gcc and g++ (and c++).

```bash
#!/bin/bash

# This is a wrapper script that adds --sysroot=/opt/vmware/glibc/sysroot to every gcc call
# It is required due to a meson bug during compiler detection that requires a valid sysroot
# VMware hacks their version of gcc to set --sysroot=/dev/null/usr/include which can never be valid

/opt/vmware/gcc/usr/bin/x86_64-vmk-linux-gnu-gcc --sysroot=/opt/vmware/glibc/sysroot $@
```

**Comment by kisdler** [comment]
BTW, where can I build and use the VMware NVDK for custom build ESXi drivers? @haata

**Comment by haata** [comment]
https://techpartnerhub.vmware.com/sdks?rel=%2Fsdks

You generally need to be granted access to this. I'm not sure what the process is for it.
